### PR TITLE
Update for Dell v5 API

### LINKED ISSUE #86: Dell v4 warranty API no longer works
Dell have released a new v5 API for warranty calls, so the module should be updated to use this instead of v4.

### PR DESCRIPTION
**Pull request by legooolas**
The Dell v4 warranty API has been replaced by v5 and so this PR updates the warranty fact to do basic lookups using this newer warranty API.

(Also notes that some newer Ubuntu versions are not supported for OpenManage but doesn't yet attempt to use the newer provided repos as of yet -- will be in a separate PR!)

Fixes #86

### CODE REVIEW

**Comment by legooolas** [comment]
Note that the parameter changes will be a breaking change, so this should be a major version change?

File: lib/facter/warranty.rb [hunk 1]
```diff
@@ -17,40 +17,72 @@ def create_dell_warranty_cache(cache)
...
 
+  # Dell v5 API requires OAuth2 setup before the call, but this just adds one minor step
+  # (ends up with the same data in the response, in a very similar format to v4 responses!)
+
   begin
-    # rescue in case dell.com is down
-    dell_api_key     = File.read("/etc/dell_api_key") # Production API key needs to be grabbed from a file
-    uri              = URI.parse("https://api.dell.com/support/assetinfo/v4/getassetwarranty/#{servicetag}?apikey=#{dell_api_key}")
+    dell_client_id     = File.read("/etc/dell_client_id")     # API v5 client_id needs to be grabbed from a file
+    dell_client_secret = File.read("/etc/dell_client_secret") # API v5 client_secret needs to be grabbed from a file
```
Comment (raphink) [inline lib/facter/warranty.rb]: I'd rather have a single file for that setting, so maybe `/etc/dell_warranty.conf` with `id` and `secret` entries in there

File: lib/facter/warranty.rb [hunk 2]
```diff
@@ -17,40 +17,72 @@ def create_dell_warranty_cache(cache)
...
+
+    # Examples using curl:
+    # 1) OAuth2 POST to get bearer token
+    #   curl -X POST --data 'grant_type=client_credentials' --data 'client_id=<id>' --data 'client_secret=<secret>'  https://apigtwb2c.us.dell.com/auth/oauth/v2/token
+    # 2) API v5 GET passing returned token to get JSON warranty info as before
+    #   curl -X GET https://apigtwb2c.us.dell.com/PROD/sbil/eapi/v5/asset-entitlements?servicetags=<tags> --header 'Authorization: Bearer <token_returned_by_first_call>'
```
Comment (raphink) [inline lib/facter/warranty.rb]: Not sure the comments are useful to keep here

File: lib/facter/warranty.rb [hunk 3]
```diff
@@ -17,40 +17,72 @@ def create_dell_warranty_cache(cache)
...
+      :grant_type    => 'client_credentials',
+      :client_id     => dell_client_id,
+      :client_secret => dell_client_secret,
+    }
+    # TODO : Ensure SSL verified!
+    oauth_res        = Net::HTTP.post_form oauth_post_uri, oauth_params
+    oauth_r          = JSON.parse(oauth_res.body)
+    bearer_token     = oauth_r['access_token']
+
+    # TODO : Fail if no result or token not found
```
Comment (raphink) [inline lib/facter/warranty.rb]: We definitely need error management. I'd return the cached value if any.

**Comment by JGodin-C2C** [comment]
Stalled for 4 (!) years. closing now.

**Pull request closed without merging**